### PR TITLE
feature(nyz&dcy): add LLM/VLM RLHF loss (PPO/GRPO/RLOO)

### PR DESCRIPTION
**Pull request by PaParaZz1**
## Description

- [x] polish and test original PPO
- [x] implement GRPO and RLOO
- [x] optimize efficiency in real LLM cases

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/rl_utils/grpo.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def grpo_policy_error(
+        data: namedtuple,
+        clip_ratio: float = 0.2,
+        beta: float = 0.1,  # Weight coefficient for KL divergence
+) -> Tuple[namedtuple, namedtuple]:
+    """Calculate the policy loss for GRPO
+    Args:
+        data (grpo_policy_data): Data containing the following fields:
```
Comment (PaParaZz1) [inline ding/rl_utils/grpo.py:15]: polish comment formats

File: ding/rl_utils/grpo.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+grpo_policy_data = namedtuple('grpo_policy_data', ['logit_new', 'logit_old', 'logit_ref', 'action', 'adv', 'weight'])
+
+
+def grpo_policy_error(
+        data: namedtuple,
+        clip_ratio: float = 0.2,
+        beta: float = 0.1,  # Weight coefficient for KL divergence
+) -> Tuple[namedtuple, namedtuple]:
+    """Calculate the policy loss for GRPO
```
Comment (PaParaZz1) [inline ding/rl_utils/grpo.py:13]: add paper link

File: ding/rl_utils/rloo.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    # Calculate additional metrics
+    metrics = {
+        'mean_ratio': ((ratio * weight).sum(dim=1) / weight.sum(dim=1)).mean().item(),
+        'mean_clipped': (ratio > (1 + clip_ratio)).float().mean().item() + (ratio <
+                                                                            (1 - clip_ratio)).float().mean().item(),
+        'mean_advantage': advantages.mean().item(),
+    }
+
+    # Create return namedtuples
+    loss_info = namedtuple('LossInfo', ['policy_loss'])(policy_loss=loss)
```
Comment (PaParaZz1) [inline ding/rl_utils/rloo.py:66]: if there is just one field, you can directly return it rather than use namedtuple

File: ding/rl_utils/rloo.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    metric_info = namedtuple('MetricInfo', list(metrics.keys()))(**metrics)
```
Comment (PaParaZz1) [inline ding/rl_utils/rloo.py:67]: you can define namedtuple at the beginning of this file

File: ding/rl_utils/grpo.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    Args:
+        data (grpo_policy_data): Data containing the following fields:
+            - logit_new: Current policy logits [B, L, V]
+            - logit_old: Old policy logits [B, L, V]
+            - logit_ref: Reference policy logits [B, L, V]
```
Comment (PaParaZz1) [inline ding/rl_utils/grpo.py:18]: B, S, N

File: ding/rl_utils/grpo.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            - action: Actions taken [B, L]
+            - adv: Advantage values [B]
+            - weight: Attention mask [B, L]
```
Comment (PaParaZz1) [inline ding/rl_utils/grpo.py:21]: use the extra `Shapes` part

File: ding/rl_utils/grpo.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        clip_ratio (float): PPO clipping ratio, default 0.2
+        beta (float): Weight coefficient for KL divergence, default 0.1
```
Comment (PaParaZz1) [inline ding/rl_utils/grpo.py:23]: add a period to the end of each sentence.

File: ding/rl_utils/grpo.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        beta: float = 0.1,  # Weight coefficient for KL divergence
+) -> Tuple[namedtuple, namedtuple]:
+    """Calculate the policy loss for GRPO
+    Args:
+        data (grpo_policy_data): Data containing the following fields:
+            - logit_new: Current policy logits [B, L, V]
+            - logit_old: Old policy logits [B, L, V]
+            - logit_ref: Reference policy logits [B, L, V]
+            - action: Actions taken [B, L]
+            - adv: Advantage values [B]
```
Comment (PaParaZz1) [inline ding/rl_utils/grpo.py:20]: [B, ]

File: ding/rl_utils/grpo.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+) -> Tuple[namedtuple, namedtuple]:
+    """
+        Overview:
+            Implementation of Generalized Reward-Conditioned Policy Optimization(	arXiv:2405.20304) .
+        Arguments:
+            - data (:obj:`namedtuple`): the grpo input data with fields shown in ``grpo_policy_data``.
+            - clip_ratio (:obj:`float`): the ppo clip ratio for the constraint of policy update, defaults to 0.2.
+            - beta (:obj:`float`): weight coefficient for KL divergence regularization, defaults to 0.1.
+        Returns:
+             - loss (:obj:`torch.FloatTensor`): the rloo policy loss, a differentiable 0-dim tensor.
```
Comment (PaParaZz1) [inline ding/rl_utils/grpo.py:55]: a redundant space

File: ding/rl_utils/grpo.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+def less_efficient_method(logits, action):
+    dist = torch.distributions.categorical.Categorical(logits=logits)
+    logp = dist.log_prob(action)
+    return logp
+
+
+def grpo_policy_error(
+        data: namedtuple,
+        logpro_cal=efficient_method,  # Method to calculate the log probabilities
```
Comment (PaParaZz1) [inline ding/rl_utils/grpo.py:43]: typing lint

File: ding/rl_utils/grpo.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        clip_ratio: float = 0.2,
+        beta: float = 0.1  # Weight coefficient for KL divergence
+) -> Tuple[namedtuple, namedtuple]:
+    """
+        Overview:
+            Implementation of Generalized Reward-Conditioned Policy Optimization(	arXiv:2405.20304) .
+        Arguments:
```
Comment (PaParaZz1) [inline ding/rl_utils/grpo.py:50]: add new arguments in comments

File: ding/rl_utils/rloo.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+def naive_method(logits, index):
+    # Calculate log probabilities for each token
+    log_prob_new = torch.log_softmax(logits, dim=-1)
+    # Get log probabilities for selected actions
+    index = index.unsqueeze(-1)  # [B, L, 1]
+    per_token_logps = torch.gather(log_prob_new, -1, index).squeeze(-1)
+    return per_token_logps
+
+
+def efficient_method(logits, index):
```
Comment (PaParaZz1) [inline ding/rl_utils/rloo.py:18]: don't repeat definition

File: ding/rl_utils/tests/test_grpo_rlhf.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    assert isinstance(logit_new.grad, torch.Tensor)
+
+    # 6. Verify metrics
+    assert 'mean_kl' in info._asdict()
+    assert 'mean_ratio' in info._asdict()
+    assert 'mean_clipped' in info._asdict()
+    assert all([np.isscalar(v) for v in info._asdict().values()])
+
+
+@pytest.mark.benchmark
```
Comment (PaParaZz1) [inline ding/rl_utils/tests/test_grpo_rlhf.py:112]: gputest

File: ding/rl_utils/tests/test_grpo_rlhf.py [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    vocab_size = 32768
+    seq_len = 1024
+    batch_size = 16
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+
+    # 生成测试数据
+    logits = torch.randn(batch_size, seq_len, vocab_size, device=device, dtype=torch.float32)
+    input_ids = torch.randint(0, vocab_size, (batch_size, seq_len), device=device)
+
+    # 预热 GPU
```
Comment (PaParaZz1) [inline ding/rl_utils/tests/test_grpo_rlhf.py:125]: English comments
